File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/MultiLayerTransformer.java
package edu.uci.ics.jung.visualization;

import edu.uci.ics.jung.visualization.transform.BidirectionalTransformer;
import edu.uci.ics.jung.visualization.transform.MutableTransformer;
import edu.uci.ics.jung.visualization.transform.shape.ShapeTransformer;
import edu.uci.ics.jung.visualization.util.ChangeEventSupport;
import java.awt.Shape;
import java.awt.geom.Point2D;

public interface MultiLayerTransformer
    extends BidirectionalTransformer, ShapeTransformer, ChangeEventSupport {

  enum Layer {
    LAYOUT,
    VIEW
  }

  void setTransformer(Layer layer, MutableTransformer Function);

  MutableTransformer getTransformer(Layer layer);

  Point2D inverseTransform(Layer layer, Point2D p);

  Point2D inverseTransform(Layer layer, double x, double y);

  Point2D transform(Layer layer, Point2D p);

  Point2D transform(Layer layer, double x, double y);

  Shape transform(Layer layer, Shape shape);

  Shape inverseTransform(Layer layer, Shape shape);

  void setToIdentity();
}

File: jung-visualization/src/main/kotlin/edu/uci/ics/jung/visualization/MultiLayerTransformer.kt
package edu.uci.ics.jung.visualization

import edu.uci.ics.jung.visualization.transform.BidirectionalTransformer
import edu.uci.ics.jung.visualization.transform.MutableTransformer
import edu.uci.ics.jung.visualization.transform.shape.ShapeTransformer
import edu.uci.ics.jung.visualization.util.ChangeEventSupport
import java.awt.Shape
import java.awt.geom.Point2D

interface MultiLayerTransformer : BidirectionalTransformer, ShapeTransformer, ChangeEventSupport {

    enum class Layer {
        LAYOUT,
        VIEW
    }

    fun setTransformer(layer: Layer, function: MutableTransformer)

    fun getTransformer(layer: Layer): MutableTransformer

    fun inverseTransform(layer: Layer, p: Point2D): Point2D?

    fun inverseTransform(layer: Layer, x: Double, y: Double): Point2D?

    fun transform(layer: Layer, p: Point2D): Point2D?

    fun transform(layer: Layer, x: Double, y: Double): Point2D?

    fun transform(layer: Layer, shape: Shape): Shape?

    fun inverseTransform(layer: Layer, shape: Shape): Shape?

    fun setToIdentity()
}
